File: BLINKER/Frontend/vehicle_me/lib/domain/models/typing_event.dart
enum Typing { start, stop }

extension TypingParsing on Typing {
  String value() {
    //returns Typing String
    return this
        .toString()
        .split(".")
        .last;
  }

  static Typing fromString(String? event) {
    //returns Typing value
    return Typing.values.firstWhere((element) =>
    element.value() == event);
  }
}
class TypingEvent {
  String? get id => _id;
  final String sender;
  final String recipient;
  final Typing event;
  String? _id;
  String? chatId;


  TypingEvent({
    required this.chatId,
    required this.sender,
    required this.recipient,
    required this.event,
  });


  TypingEvent.empty()
      : sender = '',
        recipient = '',
        event = Typing.start;


  factory TypingEvent.fromJson(Map<String, dynamic> json) {
    return TypingEvent(
      sender: json['messageID'],
      recipient: json['recipient'] ,
      event: TypingParsing.fromString(json['status']),
      chatId: json['chatId'],
    ).._id = json['_id']; //TypingEvent
  }

  Map<String, dynamic> toJson() => {
    'chatId' : chatId,
    'sender': sender,
    'recipient': recipient,
    'event': event.value(),
  };
}

File: BLINKER/Backend/Backend/app/models/typingEvent.py
from pydantic import BaseModel
from enum import Enum
from typing import Optional

class Typing(str, Enum):
    START = "start"
    STOP = "stop"

class TypingEvent(BaseModel):
    id: Optional[str] = None
    sender: str
    recipient: str
    event: Typing
    chatId: Optional[str] = None

    class Config:
        allow_population_by_field_name = True

    @classmethod
    def from_dict(cls, data):
        # Adjust the field names to match the Dart model
        if 'messageID' in data:
            data['sender'] = data.pop('messageID')
        if 'status' in data:
            data['event'] = data.pop('status')
        return cls(**data)

    def to_dict(self):
        result = self.dict(exclude_none=True)
        # Adjust the field names to match the Dart model
        if 'sender' in result:
            result['messageID'] = result.pop('sender')
        if 'event' in result:
            result['status'] = result.pop('event')
        return result